### DISCUSSION TITLE
Techniques to speed up `import-db`?

### DISCUSSION THREAD

**Discussion by pbowyer** [ddev/ddev#6591] [Q&A]
I'm finding myself dealing with large databases at the moment, and a 1.1GB `.sql.gz` file (19GB uncompressed) is taking over 22 minutes to import into DDEV, starting at 15MiB/s and by the end (presumably as rows are inserted) reporting 6-8MiB/s.

I'm dumping the source database with `mysqldump --single-transaction dbname > dbname.sql` and then gzipping. It's using extended inserts by default so is efficiently sending data in.

I'd welcome any tips, tricks or ideas to speed this up - or pointers for how to profile it to see where the time is going?

Current setup is:
```
OrbStack Version: 1.7.5 (1070500)
ddev version v1.23.4
MySQL 8.0 database image (but image doesn't seem to affect load times)
```

**Comment by rfay** [comment]
Please see
* https://github.com/ddev/ddev/pull/6411
* https://github.com/ddev/ddev/issues/6244

Interested in your feedback on that. 
Unfortunately, that PR went into v1.23.4, which is apparently what you're using. You may want to try some of the other mentioned workarounds, but this does seem like a problem with mysql 8.0. 

The most important thing you can probably do is work on a way to sanitize and slim down the db for development. that saves far more time importing than the effort in working on the utility to do it.

Certainly open to any other techniques you may find.

**Comment by douggreen** [comment]
I struggled with this for years. I have one project database that is 7.1G compressed sql. The best solution I found was what @rfay mentioned above, I finally wrote a shell script that just processed the raw sql.gz and rewrote it, removing cruft. We can't run this on our production database, as that historical data is needed, and we couldn't lock the production database for the time that it would take. But running this on our exports that developers import locally, has brought load times from 4 hours to 5 minutes. My script is highly specific to our data, so it's not sharable.

**Reply by douggreen** [reply]
FYI, my script reduces the compressed sql from 7.1G down to 254M.

**Reply by rfay** [reply]
Maybe you can at least explain the general logic of the script? Or get ChatGPT to explain it?

**Reply by douggreen** [reply]
It's a standalone php script. It reads the .sql.gz file one line at a time, looks for INSERT statements with a preg_match, parses, processes, and rewrites them, truncating certain tables. The script knows about entity and field tables, knows about data and revision tables, and also cache tables, as well as assumes the order of the columns. The bulk of our cruft was in unused revisions, so there's a lot of logic to determine which revisions aren't needed for local dev testing. But there is also one custom business related table that was huge and mostly not needed.

In a loop it starts with
```
      // Get the table name quickly, skipping all ut the preprocess tables.
      if (substr($line, 0, 6) != 'INSERT' || !preg_match('/^INSERT INTO `(\w+)`/', $line, $insert_into)) {
        continue;
      }
```
If the table is being being truncated, it just removes the INSERT. If it's not being truncated, it then parses the INSERT and rewrites it as needed. 

Parsing the line is sometimes complicated. All real parsers were too slow, so I relied on patterns. I have additional logic for the tables that have serialized data. But for most tables it's as simple as the following. 

```
function parseRows($sting) {
  $rows = explode('),(', trim($string, ')('));
  foreach ($rows as $row) {
      $value = str_getcsv($string, ',', "'");
  }
}
```

Then you can process the values clauses, remove the values that you don't need, put the row back together and write it to the trimmed sql.

I call my script dbdiet.php :) This script takes the 7.1G file and reduces it to 254M in about 5 minutes, that sql then imports in about 5 minutes, and I can then snapshot it in a couple of minutes. The total time to have a local ddev database that is restorable in a matter of a minute, now takes 15 minutes instead of half a day! 

That's about as much as I can share from this client. And while this isn't a plug for work for me, if anyone would like help implementing this, I am available for to help on a contract basis.

**Comment by rfay** [comment]
See https://ddev.com/blog/database-improvements/ for latest improvements, including significant improvement for MySQL 8.0 import speed, and a testing framework.
